File: BinarySearch/209_MinimumSizeArray/main.py
def minSubArrayLen(self, target, nums):
    """
    :type target: int
    :type nums: List[int]
    :rtype: int
    """
    ### Slide Window
    ### [2,3,1,2,4,3] --> [2,5,6,8,12,15]
    ### Search if r-l > target 
    i, res = 0, len(nums) + 1
    for j in range(len(nums)):
        target -= nums[j]
        while target <= 0:
            res = min(res, j - i + 1)
            target += nums[i]
            i += 1
    return res % (len(nums) + 1)

File: BinarySearch/209_MinimumSizeArray/main.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int n = nums.size(), ans = INT_MAX;
        vector<int> sums(n+1,0);
        for(int i = 1; i<=n;i++){
            sums[i] = sums[i-1]+nums[i-1];    
        }
        for(int i = n; i>=0 and sums[i]>=target; i--){
            int upperbound = upper_bound(sums.begin(),sums.end(),sums[i]-target)-sums.begin();
            ans = min(ans, i-upperbound+1);
        }
        return ans==INT_MAX? 0:ans;
    }
};